### PR TITLE
add support for time quantum inserts with explicit timestamps (fb-1558)

### PR DESCRIPTION
**Pull request by paddyjok**
This PR adds support for time quantum inserts with explicit timestamps.

Oh, and tests.

### CODE REVIEW

File: sql3/errors.go [hunk 1]
```diff
@@ -706,6 +707,13 @@ func NewErrInsertValueOutOfRange(line, col int, columnName string, rowNumber int
 	)
 }
 
+func NewErrUnexpectedTimeQuantumTupleLength(line, col int, columnName string, rowNumber int, badValue []interface{}, length int) error {
```
Comment (seebs) [inline sql3/errors.go:710]: Unexpected Time Quantum Tuple Length

value `%v` out of range

I don't feel like these describe the same problem?

Comment (paddyjok) [reply sql3/errors.go:710]: oops - copy pasta

File: sql3/planner/opinsert.go [hunk 2]
```diff
@@ -264,6 +265,59 @@ func (i *insertRowIter) Next(ctx context.Context) (types.Row, error) {
...
+						return nil, sql3.NewErrUnexpectedTimeQuantumTupleLength(0, 0, columnName, rowNumber+1, v, len(v))
+					}
+
+					tupleType, ok := iv.Type().(*parser.DataTypeTuple)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected tuple type '%T'", v[0])
+					}
+
+					// first member must be a timestamp or coercable as one
+					cval, err := coerceValue(tupleType.Members[0], parser.NewDataTypeTimestamp(), v[0], parser.Pos{Line: 0, Column: 0})
```
Comment (seebs) [inline sql3/planner/opinsert.go:281]: do we not have a valid position we can use here?

also... should the timestamp be the first value? my intuition would have been that the timestamp was the last value in the tuple. did we specify this anywhere? is there a thing we're doing this to be compatible with?

**Review by seebs — APPROVED** [review]
I think this looks okay, but I'm a little uncertain about the order of the tuple, because I thought we had historically done (value, timestamp) for inserting things with timestamps. Changing that isn't the end of the world but I want to at least bring it up.

File: sql3/planner/opinsert.go (hunk 2 above)
Comment (paddyjok) [reply sql3/planner/opinsert.go:281]: Completely arbitrary. If someone can point to prior art where it's the other way, happy to change it.

**Pull request merged**